File: src/py/modskel.py
from augas import *
from buffer import *
from interpreter import *
import log

# void writelog(level, msg)
# string error()
# void reconf()
# void stop()
# void post(sname, type, user)
# void delegate(sname, type, user)
# string getenv(name)
# void shutdown(sock)
# int tcpconnect(sname, host, serv, user)
# int tcplisten(sname, host, serv, user)
# void send(sock, buffer buf)
# void setrwtimer(sock, ms, flags)
# void resetrwtimer(sock, ms, flags)
# void cancelrwtimer(sock, flags)
# int settimer(sname, ms, user)
# bool resettimer(timer, ms)
# bool canceltimer(timer)

interp = Interpreter({
    "echo": lambda x: x,
    "ping": lambda: "pong"
    })

def term(sname):
    log.debug("term(): %s" % sname)

def init(sname):
    log.debug("init(): %s" % sname)
    tcplisten(sname, "0.0.0.0", getenv("session.modskel.serv"), None)

def reconf(sname):
    log.debug("reconf(): %s" % sname)

def event(sname, type, user):
    log.debug("event(): %s" % sname)

def closed(sock):
    log.info("closed(): %s" % sock)

def teardown(sock):
    log.debug("teardown(): %s" % sock)
    shutdown(sock)

def accept(sock, addr, port):
    log.info("accept(): %s" % sock)
    sock.user = Buffer()
    setrwtimer(sock, 15000, TIMRD)
    send(sock, "OK: hello\r\n")

def connected(sock, addr, port):
    log.debug("connected(): %s" % sock)

def data(sock, buf):
    log.debug("data(): %s" % sock)
    global interp
    for line in sock.user.lines(str(buf)):
        s = interp(line)
        if s == Quit:
            shutdown(sock)
        elif s != None:
            send(sock, s + "\r\n")

def rdexpire(sock, ms):
    log.debug("rdexpire(): %s" % sock)
    shutdown(sock)

def wrexpire(sock, ms):
    log.debug("wrexpire(): %s" % sock)

def expire(timer, ms):
    log.debug("expire(): %s" % timer)

File: src/cpp/augas/modskel.c
#include "augas.h"

static void
stop_(const struct augas_serv* serv)
{
    augas_writelog(AUGAS_LOGINFO, "stop_()");
}

static int
start_(struct augas_serv* serv)
{
    augas_writelog(AUGAS_LOGINFO, "start_()");
    return 0;
}

static int
reconf_(const struct augas_serv* serv)
{
    augas_writelog(AUGAS_LOGINFO, "reconf_()");
    return 0;
}

static int
event_(const struct augas_serv* serv, const char* from,
        const struct augas_event* event)
{
    augas_writelog(AUGAS_LOGINFO, "event_()");
    return 0;
}

static void
closed_(const struct augas_object* sock)
{
    augas_writelog(AUGAS_LOGINFO, "closed_()");
}

static int
teardown_(const struct augas_object* sock)
{
    augas_writelog(AUGAS_LOGINFO, "teardown_()");
    augas_shutdown(sock->id_);
    return 0;
}

static int
accept_(struct augas_object* sock, const char* addr, unsigned short port)
{
    augas_writelog(AUGAS_LOGINFO, "accept_()");
    return 0;
}

static int
connected_(struct augas_object* sock, const char* addr, unsigned short port)
{
    augas_writelog(AUGAS_LOGINFO, "connected_()");
    return 0;
}

static int
data_(const struct augas_object* sock, const char* buf, size_t size)
{
    augas_writelog(AUGAS_LOGINFO, "data_()");
    augas_send(sock->id_, buf, size);
    return 0;
}

static int
rdexpire_(const struct augas_object* sock, unsigned* ms)
{
    augas_writelog(AUGAS_LOGINFO, "rdexpire_()");
    return 0;
}

static int
wrexpire_(const struct augas_object* sock, unsigned* ms)
{
    augas_writelog(AUGAS_LOGINFO, "wrexpire_()");
    return 0;
}

static int
expire_(const struct augas_object* timer, unsigned* ms)
{
    augas_writelog(AUGAS_LOGINFO, "expire_()");
    return 0;
}

static const struct augas_module module_ = {
    stop_,
    start_,
    reconf_,
    event_,
    closed_,
    teardown_,
    accept_,
    connected_,
    data_,
    rdexpire_,
    wrexpire_,
    expire_
};

static const struct augas_module*
init_(const char* name)
{
    augas_writelog(AUGAS_LOGINFO, "init_()");
    return &module_;
}

static void
term_(void)
{
    augas_writelog(AUGAS_LOGINFO, "term_()");
}

AUGAS_MODULE(init_, term_)
